pep-0572.rst

```diff
@@ -1027,6 +1028,123 @@ To my eyes, the original form is harder to understand::
     return a
 
 
+Appendix B: Rough code translations for comprehensions
+======================================================
+
+This appendix attempts to clarify (though not specify) the rules when
+a target occurs in a comprehension or in a generator expression.
+For a number of illustrative examples we show the original code,
+containing a comprehension, and the translation, where the
+comprehension has been replaced by an equivalent generator function
+plus some scaffolding.
+
+Since ``[x for ...]`` is equivalent to ``list(x for ...)`` these
+examples all use list comprehensions without loss of generality.
+And since these examples are meant to clarify edge cases of the rules,
+they aren't trying to look like real code.
+
+Let's start with a reminder of what code is generated for a generator
+expression without assignment expression.
+
+- Original code (EXPR usually references VAR)::
+
+    def f():
+        a = [EXPR for VAR in ITERABLE]
+
+- Translation (let's not worry about name conflicts)::
+
+    def f():
+        def genexpr(iterator):
+            for VAR in iterator:
+                yield EXPR
+        a = list(genexpr(iter(ITERABLE)))
+
+Let's add a simple assignment expression.
+
+- Original code::
+
+    def f():
+        a = [TARGET := EXPR for VAR in ITERABLE]
+
+- Translation::
+
+    def f():
+        if 0:
+            TARGET = 0  # Dead code to ensure TARGET is a local variable
```

Comment:
Are typecheckers expected to ignore dead code when inferring types?

My rationale for starting to use the complex dance of:

    if 0:
        for TARGET in (): pass

Was that:

1. The `if 0:` ensures there's no actual bytecode emitted for this, it's just a symbol table pass thing for the compiler, akin to `nonlocal TARGET` or `global TARGET`
2. The `for TARGET in (): pass` ensures there's no actual assignment for a typechecker to attempt to infer anything from
3. The symtable pass does *not* ignore unreachable code, so it still marks `TARGET` as being local to `f`

Perhaps rather than the `parentlocal` I suggested on the mailing list, this appendix could benefit from an `__unboundlocal TARGET` annotation convention to indicate this state? It's akin to `TARGET: typing.Any`, but the type is still inferred from the first binding operation, not the annotation.

Reply:
Would an annotation work here? `TARGET: ...` makes the name local, while still emitting no code.

Reply:
The translated code was meant for Python's execution engine (e.g. CPython's bytecode compiler + runtime).

I wasn't actually sure what mypy does with dead code -- I think early versions still checked it, but the current version is indeed silent on Nick's construct, but I feel the target audience for this appendix are humans, and I'd have to add a paragraph to explain it.

Introducing a new notation just for this purpose would suffer the same problem: I'd have to add more text to explain it.

I considered `TARGET: Any` or `TARGET: object`, but these would break if there was another declaration or initialization of `TARGET` in the same scope (at least mypy doesn't like this). Also, `TARGET: object` would require a `cast()` call or an `isinstance()` check before `TARGET` could be passed to something with a more precise type.

So all in all I decided that `if 0: TARGET = 0` was the best solution for this appendix -- it does the right thing for CPython, and human compiler writers should understand it.

Reply:
Question:   I've never used type annotations, so have no idea.  Will type checkers complain about `TARGET = 0` if, e.g., `TARGET: list` (assuming that's how you spell "TARGET is of type list" - no idea) appears elsewhere?

I'm not clear on why we're concerned about type checkers here, but perhaps this idea (which I don't believe I mentioned before) could worm around that:

    if 0:
        del TARGET

That's enough for CPython to decide it's local too, but like Nick's `for TARGET in (): pass` says nothing from which type could be inferred.

Reply:
Nah, `if 0: del TARGET` gives an error in mypy: `Name 'TARGET' is not defined`.

It's really best not to worry about type checkers here. Type checkers aren't going to use these translations. They are going to use an internal API to ensure that a variable has local scope. (Ditto for the execution engine of course, but the fiction here is that all of us reading this PR have a decent execution engine for Python in our heads.)

Reply:
Targeting humans with these examples makes sense, but I'm not sure it's possible to avoid adding some explanatory text about their intended applicability.

How would you feel about using:

    if 0:
        TARGET = None

And then adding a bit to the appendix introduction saying something like:

> Note: the examples in this appendix are intended primarily to help compiler authors in developing tests for edge cases in their implementations. For type inference purposes, these illustrative expansions do *not* imply that assignment expression targets are always `Optional` (but they do indicate the target binding scope).

Reply:
I like Nick's suggestion just above this, and would add a bit more to clarify something else for non-compiler authors too:

> Note:  comprehensions are already implemented via synthesizing nested generator functions like those in this appendix.  The new part is adding appropriate declarations to establish the intended scope of assignment expression targets (the same scope they resolve to _as if_ the assignment were performed in the block containing the outermost comprehension).   For type inference purposes, these illustrative expansions do not imply that assignment expression targets are always `Optional` (but they do indicate the target binding scope).